gateway-js/src/config.ts

```diff
@@ -143,45 +153,90 @@ export interface ManagedGatewayConfig extends GatewayConfigBase {
   schemaConfigDeliveryEndpoint?: string;
 }
 
+// TODO(trevor:removeServiceList): migrate users to `supergraphSdl` function option
 interface ManuallyManagedServiceDefsGatewayConfig extends GatewayConfigBase {
+  // @deprecated: use `supergraphSdl` in its function form instead
   experimental_updateServiceDefinitions: Experimental_UpdateServiceDefinitions;
 }
 
-interface ManuallyManagedSupergraphSdlGatewayConfig extends GatewayConfigBase {
-  experimental_updateSupergraphSdl: Experimental_UpdateSupergraphSdl
+// TODO(trevor:removeServiceList): migrate users to `supergraphSdl` function option
+interface ExperimentalManuallyManagedSupergraphSdlGatewayConfig
+  extends GatewayConfigBase {
+  // @deprecated: use `supergraphSdl` in its function form instead
+  experimental_updateSupergraphSdl: Experimental_UpdateSupergraphSdl;
+}
+
+export function isManuallyManagedSupergraphSdlGatewayConfig(
+  config: GatewayConfig,
+): config is ManuallyManagedSupergraphSdlGatewayConfig {
+  return (
+    'supergraphSdl' in config && typeof config.supergraphSdl === 'function'
+  );
+}
+
+export interface SupergraphSdlUpdateFunction {
+  (updatedSupergraphSdl: string): Promise<void>
+}
+export interface SupergraphSdlHookOptions {
+  update: SupergraphSdlUpdateFunction;
+}
+
+export type SupergraphSdlHookReturn = Promise<{
+  supergraphSdl: string;
+  cleanup?: () => Promise<void>;
+}>;
+
+export interface SupergraphSdlHook {
```

Comment:
(a) isn't there a syntax for declaring function types that's simpler?
(b) I feel like it reads better for the Promise<> to go here rather than in the SupergraphSdlHookReturn type, just like you do for SupergraphSdlUpdateFunction.

Reply:
(or maybe `supergraphSdl: (options: SupergraphSdlHookOptions) => Promise<SupergraphSdlHookReturn>` and drop this type? idk)

Reply:
Limited the exported types to those which will be useful to implementors (i.e. I need to type my `let update: UpdateFn` since it's assigned later in the callback.

Resolved in c64be0fb